### PR TITLE
Orchestrator refactor

### PR DESCRIPTION
**Pull request by SweeXordious**

### CODE REVIEW

File: x/qgb/orchestrator/orchestrator.go [hunk 1]
```diff
@@ -4,95 +4,440 @@ import (
...
+		wg.Wait()
+		return
+	}
+}
+
+func (orch Orchestrator) startNewEventsListener(ctx context.Context, queue chan<- uint64, signalChan <-chan struct{}) error {
+	results, err := orch.querier.SubscribeEvents(ctx, "attestation-changes", fmt.Sprintf("%s.%s='%s'", types.EventTypeAttestationRequest, sdk.AttributeKeyModule, types.ModuleName))
+	if err != nil {
+		panic(err)
+	}
```
Comment (evan-forbes) [inline x/qgb/orchestrator/orchestrator.go:119]: why not bubble this up?

**Review by evan-forbes — COMMENTED** [review]
some minor nits, but overall a big :+1: 

probably about as clean as we're going to get! anything smaller and I think we're treading into over abstraction

File: x/qgb/orchestrator/orchestrator.go [hunk 2]
```diff
@@ -4,95 +4,440 @@ import (
...
+	}()
+
+	// FIXME should we add  another go routine that keep checking if all the attestations
+	// were signed every 10min for example?
+
+	select {
+	case <-withCancel.Done():
+		wg.Wait()
+		return
+	}
```
[the remark this answers is not in the document]
Comment (SweeXordious) [reply x/qgb/orchestrator/orchestrator.go:112]: agreed

File: x/qgb/orchestrator/orchestrator_command.go [hunk 3]
```diff
@@ -29,122 +23,70 @@ func OrchestratorCmd() *cobra.Command {
...
+	var sigCh = make(chan os.Signal)
+
+	signal.Notify(sigCh, syscall.SIGTERM)
+	signal.Notify(sigCh, syscall.SIGINT)
+
+	go func() {
+		sig := <-sigCh
+		logger.Info("caught signal; shutting down...", "signal", sig.String())
+		cancel()
+	}()
```
[the remark this answers is not in the document]
Comment (SweeXordious) [reply x/qgb/orchestrator/orchestrator_command.go:91]: I think the name `trapSignal` makes sense. If you have a suggestion on a specific name, that would be good. I can't think of any better name

File: x/qgb/orchestrator/orchestrator.go [hunk 4]
```diff
@@ -4,95 +4,440 @@ import (
...
+		err := orch.startNewEventsListener(withCancel, noncesQueue, signalChan)
 		if err != nil {
-			oc.logger.Error(fmt.Sprintf("valset nonce %d: %s", valset.Nonce, err.Error()))
-			continue
+			orch.logger.Error("error listening to new attestations", "err", err)
+			cancel()
 		}
+		orch.logger.Error("stopping listening to new attestations")
+		wg.Done()
+	}()
```
[the remark this answers is not in the document]
Comment (SweeXordious) [reply x/qgb/orchestrator/orchestrator.go:92]: there is a  defer in here: 
https://github.com/celestiaorg/celestia-app/blob/4406b88156d543b7a223c48de2e9a3274c054dde/x/qgb/orchestrator/orchestrator.go#L67

Is that what you mean?

[the remark this answers is not in the document]
Comment (SweeXordious) [reply x/qgb/orchestrator/orchestrator.go:92]: Aaah, probably would need to move it up a couple  lines to be clear

File: x/qgb/orchestrator/orchestrator.go (hunk 1 above)
Comment (SweeXordious) [reply x/qgb/orchestrator/orchestrator.go:119]: I missed it  :D  my bad

File: x/qgb/orchestrator/orchestrator.go [hunk 5]
```diff
@@ -4,95 +4,436 @@ import (
...
+		return "", err
+	}
+
+	// FIXME sdktypestx.BroadcastMode_BROADCAST_MODE_BLOCK waits for a block to be minted containing
+	// the transaction to continue. This makes the orchestrator slow to catchup.
+	// It would be better to just send the transaction. Then, another job would keep an eye
+	// if the transaction was included. If not, retry it. But this would mean we should increment ourselves
+	// the sequence number after each broadcasted transaction.
+	// We can also use BroadcastMode_BROADCAST_MODE_SYNC but it will also fail due to a non incremented
+	// sequence number.
```
Comment (SweeXordious) [inline x/qgb/orchestrator/orchestrator.go:362]: @evan-forbes please when you have time,  let me know what you think about this,

File: x/qgb/orchestrator/orchestrator.go [hunk 6]
```diff
@@ -4,95 +4,436 @@ import (
...
+
+type Retrier interface {
+	Retry(nonce uint64, ctx context.Context, retryMethod func(context.Context, uint64) error) error
+	RetryThenFail(nonce uint64, ctx context.Context, retryMethod func(context.Context, uint64) error)
+}
+
+func (r retrier) Retry(nonce uint64, ctx context.Context, retryMethod func(context.Context, uint64) error) error {
+	var err error
+	for i := 0; i <= r.retriesNumber; i++ {
+		// We can implement some exponential backoff in here
```
Comment (SweeXordious) [inline x/qgb/orchestrator/orchestrator.go:399]: @evan-forbes do we want to have this in this PR? or we leave it for later?

File: x/qgb/orchestrator/orchestrator.go [hunk 7]
```diff
@@ -4,95 +4,436 @@ import (
...
-			oc.logger.Error(fmt.Sprintf("valset nonce %d: %s", valset.Nonce, err.Error()))
-			continue
+			orch.logger.Error("error processing attestations", "err", err)
+			cancel()
+		}
+		orch.logger.Error("stopping processing attestations")
+	}()
+
+	// FIXME should we add  another go routine that keep checking if all the attestations
+	// were signed every 10min for example?
```
Comment (SweeXordious) [inline x/qgb/orchestrator/orchestrator.go:106]: @evan-forbes What do you think about this?

File: x/qgb/orchestrator/orchestrator.go [hunk 8]
```diff
@@ -4,95 +4,436 @@ import (
...
+	"strconv"
+	"sync"
+	"time"
 )
 
-type orchestrator struct {
-	logger log.Logger
-	// TODO this will change once we have the worker pool pattern
+type Orchestrator struct {
+	logger tmlog.Logger // maybe use a more general interface
```
Comment (SweeXordious) [inline x/qgb/orchestrator/orchestrator.go:26]: @evan-forbes should we use a more generic logger interface? tendermint is quite opinionated, no?

**Review by SweeXordious — COMMENTED** [review]
Questions :D

**Comment by SweeXordious** [comment]
@evan-forbes ready for final review, after you're back. Also, please check the comments up.

File: x/qgb/orchestrator/orchestrator.go (hunk 5 above)
Comment (evan-forbes) [reply x/qgb/orchestrator/orchestrator.go:399]: we might want to create larger txs with many sdk.Messages and still use broadcast mode sync just for simplicity, but I can definitely see where a more async approach would be desirable

File: x/qgb/orchestrator/orchestrator.go (hunk 6 above)
Comment (evan-forbes) [reply x/qgb/orchestrator/orchestrator.go:436]: we can leave it for later

File: x/qgb/orchestrator/orchestrator.go (hunk 8 above)
Comment (evan-forbes) [reply x/qgb/orchestrator/orchestrator.go]: the reasoning behind using the tendermint logger was just to stay consistent with what is used in the app. While I certainly agree that a different logger would probably fit our case better, using two different loggers in the same repo sounds like not a good idea. If we end up moving the orchestrator/relayer out of this repo, then I'd feel a lot better about switching it.

File: x/qgb/orchestrator/orchestrator.go (hunk 7 above)
Comment (evan-forbes) [reply x/qgb/orchestrator/orchestrator.go:118]: perhaps, yeah. If this is something that can wait, then let's wait

**Review by evan-forbes — APPROVED** [review]

File: x/qgb/orchestrator/orchestrator.go (hunk 5 above)
Comment (SweeXordious) [reply x/qgb/orchestrator/orchestrator.go:399]: Should we create an issue for that?

File: x/qgb/orchestrator/orchestrator.go (hunk 7 above)
Comment (SweeXordious) [reply x/qgb/orchestrator/orchestrator.go:118]: Let's leave it for later then. I will create an issue an issue.

File: x/qgb/orchestrator/orchestrator.go (hunk 8 above)
Comment (SweeXordious) [reply x/qgb/orchestrator/orchestrator.go]: Understandable :+1:

**Comment by SweeXordious** [comment]
merging :D

**Pull request merged**